### PR TITLE
Add schema aliases

### PR DESCRIPTION
**Pull request by dstockwell**

### CODE REVIEW

**Comment by shans** [comment]
What's the use case for these? Is it just being able to host definitions which disagree on fields but have the same name(s) in the same schema?

**Comment by dstockwell** [comment]
> What's the use case for these?

We discussed that while the use case was primarily serialization, that was incorrect as serialization will typically use inline schemas.

Aliasing as a typedef is still useful for brevity in some cases. Proposal:

```
schema Tesla Car
  Number doors
  Boolean bioweaponDefenceMode
  
typedef schema * extends Tesla as MyTesla
// equivalent to
typedef schema Tesla Car as MyTesla
  Number doors
  Boolean bioweaponDefenceMode

particle ...
  // Following 3 are equivalent
  // 1. expands to the full typedef
  in MyTesla ...
  // 2. types inferred from `Tesla` schema.
  in Tesla Car {doors, bioweaponDefenceMode}
  // 3. in both cases, serialized as
  in Tesla Car {Number doors, Boolean bioweaponDefenceMode} ...

  // Invalid
  in Tesla ...

  // TBD valid due to `create` - `Tesla` is connected to the concrete schema
  in/out/create Tesla ...
```

**Comment by dstockwell** [comment]
For now I made it invalid to infer a field type from an alias: `in MyTesla {doors}` will fail, but `in MyTesla {Number doors}` will confusingly create a schema with names: `['MyTesla']` not `['Tesla', 'Car']`.

Some options:
1. We could make it invalid to use the same name for both alias and schema name. `in MyTesla {doors}` will fail with `Attempt to use alias 'MyTesla' as schema name`.
2. We could make aliases expand names. `in MyTesla {doors}` becomes `in Tesla Car {Number doors}`. But this is somewhat confusing behavior as schemas (by design to encourage less specific types) will not do this.
3. In combination with 1. We could make the inference a bit smarter:
    ```
    alias schema Tesla Car as MySchema
      Number doors
    particle ...
      in Car {doors}
    ```
    Even though `schema Car` is not defined here, we can look at all the schemas/aliases that comprise `Car` and if the type for `doors` aligns, use that.

**Comment by shans** [comment]
I don't understand why 2 is confusing. Do you want me to review this as is, or talk about where we're going with this first?

**Comment by dstockwell** [comment]
After further discussion, schema aliases now always expand. We'll add support for parsing inline schemas as `[Name / Field]+` later.

```
schema Foo
  Number fooField

alias schema Tesla Car as MyTesla
  Boolean bioweaponDefenceMode

particle ...
  in MyTesla Foo {fooField}
  // Can be thought of as:
  in Tesla Car {Boolean bioweaponDefenceMode} Foo {Number fooField}
  // Normalizes to:
  in Tesla Car Foo {Boolean bioweaponDefenceMode, Number fooField} 
```

File: runtime/test/manifest-test.js [hunk 1]
```diff
@@ -1218,6 +1218,42 @@ resource SomeName
...
+        particle P in 'p.js'
+          P(in Thing1 Thing2 {} param)
+      `);
+      assert.fail();
+    } catch (e) {
+      assert.include(e.message, 'Could not merge schema aliases');
+    }
+  });
+
   it('can relate inline schemas to generic connections', async () => {
```
Comment (shans) [inline runtime/test/manifest-test.js]: maybe check that something like
```
alias schema Name 1 as Thing1
  Text field1
particle P in 'p.js'
  P(in Thing1 { Number field1})
```

fails too?

Comment (dstockwell) [reply runtime/test/manifest-test.js]: Done!

**Pull request merged**